### PR TITLE
feat: dump yaml including signature for flex flow run snapshot

### PR DESCRIPTION
**Pull request by elliotzh**
# Description

Update the yaml file in run snapshot with updated signature

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/_utils/flow_utils.py [hunk 1]
```diff
@@ -321,3 +323,157 @@ def parse_variant(variant: str) -> Tuple[str, str]:
             message=str(error),
             error=error,
         )
+
+
+def format_signature_type(flow_meta):
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/_utils/flow_utils.py:328]: may put all signature related into another file

Comment (wangchao1230) [reply src/promptflow-core/promptflow/_utils/flow_utils.py:328]: shall this be in devkit package?

File: src/promptflow-devkit/tests/sdk_cli_test/e2etests/test_flow_run.py [hunk 2]
```diff
@@ -1281,6 +1281,27 @@ def test_eager_flow_run_without_yaml(self, pf):
         # the YAML file will not exist in user's folder
         assert not Path(f"{EAGER_FLOWS_DIR}/simple_without_yaml/flow.flex.yaml").exists()
 
+    def test_flex_flow_run_update_signature(self, pf):
```
Comment (0mza987) [inline src/promptflow-devkit/tests/sdk_cli_test/e2etests/test_flow_run.py:1284]: Should also cover flex flow cases:
- With `flow.flex.yaml`
- Without `flow.flex.yaml`

Actually I would recomend add the check logic to existing test case?
image

image

File: src/promptflow-core/promptflow/_utils/flow_utils.py (hunk 1 above)
Comment (elliotzh) [reply src/promptflow-core/promptflow/_utils/flow_utils.py]: moved to devkit

File: src/promptflow-devkit/tests/sdk_cli_test/e2etests/test_flow_run.py (hunk 2 above)
Comment (elliotzh) [reply src/promptflow-devkit/tests/sdk_cli_test/e2etests/test_flow_run.py]: Code path is the same. Directly appending extra check after original tests will bring duplicate code. Used `pytest.mark.parameterize` to merge some related tests.

File: src/promptflow-devkit/tests/sdk_cli_test/e2etests/test_flow_run.py [hunk 3]
```diff
@@ -1268,73 +1269,117 @@ def test_flow_with_nan_inf_metrics(self, pf: PFClient, monkeypatch) -> None:
...
-            flow=my_async_entry,
-            data=f"{DATAS_DIR}/simple_eager_flow_data.jsonl",
-            # set code folder to avoid snapshot too big
-            code=f"{EAGER_FLOWS_DIR}/multiple_entries",
-            column_mapping={"input2": "${data.input_val}"},
-        )
-        assert run.status == "Completed"
-        assert "error" not in run._to_dict()
+        yaml_dict = load_yaml(local_storage._dag_path)
+        assert yaml_dict == expected_snapshot_yaml
```
Comment (0mza987) [inline src/promptflow-devkit/tests/sdk_cli_test/e2etests/test_flow_run.py:1376]: Add check for at least one non-flex run?

Comment (elliotzh) [reply src/promptflow-devkit/tests/sdk_cli_test/e2etests/test_flow_run.py:1376]: this is flex only feature?

**Review by 0mza987 — APPROVED** [review]

**Review by 0mza987 — APPROVED** [review]
